File: AoC2020-Python/day03-toboggan-trajectory.py
### ADVENT OF CODE 2020 DAY 3 TOBOGGAN TRAJECTORY ###


import helper.helper as helper
import math


## * calculate how many trees you would hit on the puzzle_input given a down and right value
def find_trees(puzzle_input, right: int, down: int):
    # get width of one input line (they are all the same width)
    hill_width = len(puzzle_input[0])

    # count trees for given slope
    start_position = 0
    trees = 0
    for element in puzzle_input[::down]:  # :: is slice operator to skip by down value
        # modulus can be used for wrap around functionality
        if element[(start_position % hill_width)] == "#":
            trees += 1
        start_position += right

    return trees


## * main function
def main():
    # print title
    helper.print_title("2020", "3", "Toboggan Trajectory")

    # get puzzle input
    puzzle_input = helper.load_puzzle_input("inputs/day03/pi.txt")

    # solve puzzle part 1 and 2 in list
    tree_count_list = []
    tree_count_list.append(find_trees(puzzle_input, 1, 1))
    tree_count_list.append(find_trees(puzzle_input, 3, 1))
    tree_count_list.append(find_trees(puzzle_input, 5, 1))
    tree_count_list.append(find_trees(puzzle_input, 7, 1))
    tree_count_list.append(find_trees(puzzle_input, 1, 2))

    # part 1 print QA
    answer = tree_count_list[1]
    helper.print_question_answer(
        "Starting at the top-left corner of your map and following a slope of right 3 and down 1, how many trees would you encounter? ",
        answer,
    )

    # solve part 2 and print QA
    answer = str(math.prod(tree_count_list))
    helper.print_question_answer(
        "What do you get if you multiply together the number of trees encountered on each of the listed slopes? ",
        answer,
    )


if __name__ == "__main__":
    main()

File: AoC2020-C++/day03-toboggan-trajectory.cpp
// * ADVENT OF CODE 2020 - DAY 3 - TOBOGGAN TRAJECTORY *

#include "helper/helper.h"

// function to find trees you would hit on given right and down values
int findTrees(std::vector<std::string> hillVector, int right, int down) {
    // get width of one of the input lines since they are all the same width
    int hillWidth = hillVector[0].length() - 1;  // remove one for the new line character

    // count trees starting at the top left of the hill using given right and down
    int startPosition = 0;
    int trees = 0;
    for (std::size_t i = 0; i < hillVector.size(); i = i + down) {
        std::string hillLine = hillVector[i];
        if (hillLine[startPosition % hillWidth] == '#') {
            trees++;
        }
        startPosition = startPosition + right;
    }

    return trees;
}

int main() {
    // print title
    helper::printTitle("2020", "3", "Toboggan Trajectory");

    // load puzzle input
    std::vector<std::string> puzzleVector = helper::loadPuzzleInputString("inputs/day03/pi.txt");

    // solve part 1
    int answer = findTrees(puzzleVector, 3, 1);

    // print question and answer
    helper::printQuestionAnswer("\nStarting at the top-left corner of your map and following a slope of right 3 and down 1, how many trees would you encounter? ", std::to_string(answer));

    // solve part 2
    int slope1 = findTrees(puzzleVector, 1, 1);
    int slope2 = findTrees(puzzleVector, 3, 1);
    int slope3 = findTrees(puzzleVector, 5, 1);
    int slope4 = findTrees(puzzleVector, 7, 1);
    int slope5 = findTrees(puzzleVector, 1, 2);

    answer = slope1 * slope2 * slope3 * slope4 * slope5;

    // print question and answer
    helper::printQuestionAnswer("What do you get if you multiply together the number of trees encountered on each of the listed slopes? ", std::to_string(answer));

    return 0;
}